token-lending/program/src/state.rs

```
@@ -33,6 +33,57 @@ pub struct LendingMarket {
     pub token_program_id: Pubkey,
 }
 
+/// Additional fee information on a reserve
+///
+/// These exist separately from interest accrual fees, and are specifically for
+/// the program owner and frontend host.  The fees are paid out as a percentage
+/// of collateral token amounts during repayments and liquidations.
+#[derive(Clone, Copy, Debug, Default, PartialEq)]
+pub struct ReserveFees {
+    /// Fee assessed on `BorrowReserveLiquidity`, expressed as a Wad.
+    /// Must be between 0 and 10^18, such that 10^18 = 1.  A few examples for
+    /// clarity:
+    /// 1% = 10_000_000_000_000_000
+    /// 0.01% (1 basis point) = 100_000_000_000_000
+    /// 0.00001% (Aave borrow fee) = 100_000_000_000
+    pub borrow_fee_wad: u64,
+    /// Amount of fee going to host account, if provided in liquidate and repay
+    pub host_fee_percentage: u8,
+}
+
+impl ReserveFees {
+    /// Calculate the owner and host fees on borrow
+    pub fn calculate_borrow_fees(
+        &self,
+        collateral_amount: u64,
+        host_pubkey: &Pubkey,
+    ) -> (u64, u64) {
+        let borrow_fee_rate = Rate::new(self.borrow_fee_wad, SCALE);
+        let host_fee_rate = Rate::from_percent(self.host_fee_percentage);
+        if borrow_fee_rate > Rate::zero() && collateral_amount > 0 {
+            let need_to_assess_host_fee =
+                host_fee_rate > Rate::zero() && *host_pubkey != Pubkey::default();
+            let minimum_fee = if need_to_assess_host_fee {
+                2 // 1 token to owner, 1 to host
+            } else {
+                1 // 1 token to owner, nothing else
+            };
+            let borrow_fee = std::cmp::max(
```

I think we should return an error if `borrow_fee >= collateral_amount`